# Experiment 1: Batch Normalization

Goal:

Determine whether Batch Normalization improves
generalization compared to the baseline CNN.

Changes from baseline:

- Added Batch Normalization after each Conv layer.

Everything else remains identical.

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [2]:
import numpy as np
import tensorflow as tf
import matplotlib.pyplot as plt

from tensorflow.keras.models import Sequential

from tensorflow.keras.layers import (
    Conv2D,
    MaxPooling2D,
    Flatten,
    Dense,
    Dropout,
    BatchNormalization
)

In [3]:
PROCESSED_PATH = "/content/drive/MyDrive/emotion_sentiment_fusion_detector/data/processed"

X_train = np.load(f"{PROCESSED_PATH}/X_train.npy")
X_val = np.load(f"{PROCESSED_PATH}/X_val.npy")
X_test = np.load(f"{PROCESSED_PATH}/X_test.npy")

y_train = np.load(f"{PROCESSED_PATH}/y_train.npy")
y_val = np.load(f"{PROCESSED_PATH}/y_val.npy")
y_test = np.load(f"{PROCESSED_PATH}/y_test.npy")

# Batch Normalization

Batch Normalization normalizes activations during training.

Benefits:

- Faster training
- More stable gradients
- Better generalization
- Reduced sensitivity to initialization

In [4]:
model = Sequential()

# BLOCK 1

model.add(
    Conv2D(
        32,
        (3,3),
        activation="relu",
        input_shape=(48,48,1)
    )
)

model.add(BatchNormalization())

model.add(
    MaxPooling2D((2,2))
)

# BLOCK 2

model.add(
    Conv2D(
        64,
        (3,3),
        activation="relu"
    )
)

model.add(BatchNormalization())

model.add(
    MaxPooling2D((2,2))
)

# BLOCK 3

model.add(
    Conv2D(
        128,
        (3,3),
        activation="relu"
    )
)

model.add(BatchNormalization())

model.add(
    MaxPooling2D((2,2))
)

model.add(Flatten())

model.add(
    Dense(
        128,
        activation="relu"
    )
)

model.add(
    Dropout(0.5)
)

model.add(
    Dense(
        7,
        activation="softmax"
    )
)

/usr/local/lib/python3.12/dist-packages/keras/src/layers/convolutional/base_conv.py:113: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [6]:
model.compile(
    optimizer="adam",
    loss="categorical_crossentropy",
    metrics=["accuracy"]
)

In [7]:
history = model.fit(
    X_train,
    y_train,
    validation_data=(X_val, y_val),
    epochs=20,
    batch_size=64
)

Epoch 1/20
359/359 ━━━━━━━━━━━━━━━━━━━━ 121s 329ms/step - accuracy: 0.3084 - loss: 1.7615 - val_accuracy: 0.2212 - val_loss: 1.8341
Epoch 2/20
359/359 ━━━━━━━━━━━━━━━━━━━━ 118s 330ms/step - accuracy: 0.3883 - loss: 1.5526 - val_accuracy: 0.4119 - val_loss: 1.4905
Epoch 3/20
359/359 ━━━━━━━━━━━━━━━━━━━━ 118s 330ms/step - accuracy: 0.4296 - loss: 1.4608 - val_accuracy: 0.4183 - val_loss: 1.4593
Epoch 4/20
359/359 ━━━━━━━━━━━━━━━━━━━━ 142s 330ms/step - accuracy: 0.4641 - loss: 1.3800 - val_accuracy: 0.4511 - val_loss: 1.3958
Epoch 5/20
359/359 ━━━━━━━━━━━━━━━━━━━━ 121s 337ms/step - accuracy: 0.4890 - loss: 1.3204 - val_accuracy: 0.4803 - val_loss: 1.3451
Epoch 6/20
359/359 ━━━━━━━━━━━━━━━━━━━━ 118s 328ms/step - accuracy: 0.5070 - loss: 1.2681 - val_accuracy: 0.4871 - val_loss: 1.3187
Epoch 7/20
359/359 ━━━━━━━━━━━━━━━━━━━━ 145s 337ms/step - accuracy: 0.5291 - loss: 1.2150 - val_accuracy: 0.4716 - val_loss: 1.3514
Epoch 8/20
359/359 ━━━━━━━━━━━━━━━━━━━━ 139s 329ms/step - accuracy: 0.5453 -

In [8]:
test_loss, test_acc = model.evaluate(
    X_test,
    y_test
)

print(test_acc)

225/225 ━━━━━━━━━━━━━━━━━━━━ 9s 39ms/step - accuracy: 0.5517 - loss: 1.6081
0.5516856908798218


## Experiment 1 — Batch Normalization

Changes:
- Added BatchNormalization after each Conv layer

Results:

Baseline:
- Validation Accuracy: 56.0%
- Test Accuracy: 56.6%

BatchNorm:
- Validation Accuracy: 54.6%
- Test Accuracy: 55.2%

Conclusion:
Batch normalization improved training accuracy but did not improve generalization on FER2013 with the current architecture.